ws-server/src/server.rs

```diff
@@ -21,477 +21,201 @@
-	/// Registers a notification method name towards the server.
-	///
-	/// Clients will then be able to call this method.
-	/// The returned object allows you to process incoming notifications.
-	///
-	/// If `allow_losses` is true, then the server is allowed to drop notifications if the
-	/// notifications handler (i.e. the code that uses [`RegisteredNotifications`]) is too slow
-	/// to process notifications.
-	///
-	/// Returns an error if the method name was already registered.
-	pub fn register_notification(
-		&self,
-		method_name: String,
-		allow_losses: bool,
-	) -> Result<RegisteredNotification, Error> {
-		if !self.registered_methods.lock().insert(method_name.clone()) {
-			return Err(Error::MethodAlreadyRegistered(method_name));
-		}
+		{
+			let subscribers = subscribers.clone();
+			self.methods.insert(
+				subscribe_method_name,
+				Box::new(move |id, _, tx, conn| {
+					let sub_id = {
+						const JS_NUM_MASK: u64 = !0 >> 11;
 
-		log::trace!("[frontend]: register_notification={}", method_name);
-		let (tx, rx) = mpsc::channel(32);
+						let sub_id = rand::random::<u64>() & JS_NUM_MASK;
 
-		self.to_back
-			.unbounded_send(FrontToBack::RegisterNotifications { name: method_name, handler: tx, allow_losses })
-			.map_err(|e| Error::Internal(e.into_send_error()))?;
+						subscribers.lock().insert((conn, sub_id), tx.clone());
 
-		Ok(RegisteredNotification { queries_rx: rx })
-	}
+						sub_id
+					};
 
-	/// Registers a method towards the server.
-	///
-	/// Clients will then be able to call this method.
-	/// The returned object allows you to handle incoming requests.
-	///
-	/// Contrary to [`register_notifications`](Server::register_notifications), there is no
-	/// `allow_losses` parameter here. If the handler is too slow to process requests, then the
-	/// server automatically returns an "internal error" to the client.
-	///
-	/// Returns an error if the method name was already registered.
-	pub fn register_method(&self, method_name: String) -> Result<RegisteredMethod, Error> {
-		if !self.registered_methods.lock().insert(method_name.clone()) {
-			return Err(Error::MethodAlreadyRegistered(method_name));
-		}
+					let json = serde_json::to_string(&JsonRpcResponse { jsonrpc: TwoPointZero, id, result: sub_id })?;
 
-		log::trace!("[frontend]: register_method={}", method_name);
-		let (tx, rx) = mpsc::channel(32);
+					tx.send(json).map_err(Into::into)
+				}),
+			);
+		}
 
-		self.to_back
-			.unbounded_send(FrontToBack::RegisterMethod { name: method_name, handler: tx })
-			.map_err(|e| Error::Internal(e.into_send_error()))?;
+		{
+			let subscribers = subscribers.clone();
+			self.methods.insert(
+				unsubscribe_method_name,
+				Box::new(move |id, params, tx, conn| {
+					let [sub_id]: [u64; 1] = serde_json::from_str(params)?;
+
+					subscribers.lock().remove(&(conn, sub_id));
+
+					let json = serde_json::to_string(&JsonRpcResponse {
+						jsonrpc: TwoPointZero,
+						id,
+						result: SubscriptionId::Num(0),
```

Comment:
this is not intended to be hardcoded?

Reply:
https://github.com/paritytech/jsonrpsee/blob/master/ws-client/src/client.rs#L399, in the client code we just send it out and don't care about the result but perhaps we should ^^